File: Codes/Level 2/Trees/BinaryTreeLevelOrderTraversal2Reverse.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> ans;
        if(root == NULL)    return ans;
        deque<TreeNode*> q;
        q.push_back(root);
        while(q.size()){
            int k = q.size();
            vector<int> currAns;
            while(k--){
                TreeNode* curr = q.front();
                q.pop_front();
                if(curr->left != NULL)   q.push_back(curr->left);
                if(curr->right != NULL)  q.push_back(curr->right);
                currAns.push_back(curr->val);
            }
            ans.push_back(currAns);
        }
        for(int i=0; i<ans.size()/2; i++){
            vector<int> temp = ans[i];
            ans[i] = ans[ans.size() - 1 - i];
            ans[ans.size() - 1 - i] = temp;
        }
        return ans;
    }
};

File: Codes/Level 2/Trees/BinaryTreeLevelOrderTraversal2Reverse.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if(root == null)    return ans;
        Deque<TreeNode> q = new ArrayDeque<>();
        q.addLast(root);
        while(q.size() > 0){
            int k = q.size();
            List<Integer> currAns = new ArrayList<>();
            while(k-- > 0){
                TreeNode curr = q.getFirst();
                q.removeFirst();
                if(curr.left != null)   q.addLast(curr.left);
                if(curr.right != null)  q.addLast(curr.right);
                currAns.add(curr.val);
            }
            ans.add(currAns);
        }
        List<List<Integer>> res = new ArrayList<>();
        for(int i = ans.size() - 1; i >= 0; i--){
            res.add(ans.get(i));
        }
        return res;
    }
}